File: php/bench/bench.php
<?php
// Performance bench for the PHP port. Emits one JSON line per
// build/bench/README.md; diagnostics go to stderr.
require_once __DIR__ . '/../src/Struct.php';

use Voxgig\Struct\Struct;

function envi($k, $d) {
    $v = getenv($k);
    return ($v !== false && ctype_digit($v)) ? intval($v) : $d;
}

$W = envi('BENCH_WIDTH', 5);
$D = envi('BENCH_DEPTH', 6);
$WARM = envi('BENCH_WARMUP', 3);
$RUNS = envi('BENCH_RUNS', 21);
$GP = envi('BENCH_GETPATH_ITERS', 50000);

function build($w, $d, $leaf) {
    if ($d == 0) return $leaf;
    $o = [];
    for ($i = 0; $i < $w; $i++) $o['k' . $i] = build($w, $d - 1, $leaf);
    return $o;
}

function nodecount($w, $d) {
    $n = 0; $p = 1;
    for ($i = 0; $i <= $d; $i++) { $n += $p; $p *= $w; }
    return $n;
}

$sink = 0;

function measure($warm, $runs, $fn) {
    for ($i = 0; $i < $warm; $i++) $fn();
    $t = [];
    for ($r = 0; $r < $runs; $r++) {
        $a = hrtime(true);
        $fn();
        $b = hrtime(true);
        $t[] = ($b - $a) / 1e6;
    }
    sort($t);
    return ['min_ms' => $t[0], 'median_ms' => $t[intdiv(count($t), 2)],
            'mean_ms' => array_sum($t) / count($t)];
}

$tree = build($W, $D, 0);
$nodes = nodecount($W, $D);
$treeA = build($W, $D, 1);
$treeB = build($W, $D, 2);
$path = implode('.', array_fill(0, $D, 'k0'));
$cb = function ($key, $val, $parent, $p) use (&$sink) { $sink += count($p); return $val; };

$specs = [
    ['clone', $nodes, function () use (&$sink, $tree) { $sink += Struct::clone($tree) ? 1 : 0; }],
    ['walk', $nodes, function () use ($tree, $cb) { Struct::walk($tree, $cb); }],
    ['merge', $nodes, function () use (&$sink, $treeA, $treeB) { $sink += Struct::merge([$treeA, $treeB]) ? 1 : 0; }],
    ['stringify', $nodes, function () use (&$sink, $tree) { $sink += strlen(Struct::stringify($tree)); }],
    ['getpath', $GP, function () use (&$sink, $tree, $path, $GP) {
        $s = 0;
        for ($i = 0; $i < $GP; $i++) $s += Struct::getpath($tree, $path) === 0 ? 1 : 0;
        $sink += $s;
    }],
];

$ops = [];
foreach ($specs as [$op, $uc, $fn]) {
    $ops[] = array_merge(['op' => $op, 'runs' => $RUNS, 'unit_count' => $uc],
                         measure($WARM, $RUNS, $fn));
}

fwrite(STDERR, "php: sink=$sink\n");
echo json_encode([
    'lang' => 'php',
    'runtime' => 'php ' . PHP_VERSION,
    'nodes' => $nodes,
    'params' => ['width' => $W, 'depth' => $D, 'warmup' => $WARM,
                 'runs' => $RUNS, 'getpath_iters' => $GP],
    'ops' => $ops,
]), "\n";

File: lua/bench/bench.lua
-- Performance bench for the Lua port. Emits one JSON line per
-- build/bench/README.md; diagnostics go to stderr. Run from the lua/ dir.
package.path = package.path .. ";./src/?.lua"
local struct = require("struct")

local function envi(k, d)
  local v = os.getenv(k)
  if v and v:match("^%d+$") then return tonumber(v) end
  return d
end

local W = envi("BENCH_WIDTH", 5)
local D = envi("BENCH_DEPTH", 6)
local WARM = envi("BENCH_WARMUP", 3)
local RUNS = envi("BENCH_RUNS", 21)
local GP = envi("BENCH_GETPATH_ITERS", 2000)

local function build(w, d, leaf)
  if d == 0 then return leaf end
  local o = setmetatable({}, { __jsontype = "object" })
  for i = 0, w - 1 do o["k" .. i] = build(w, d - 1, leaf) end
  return o
end

local function nodecount(w, d)
  local n, p = 0, 1
  for _ = 0, d do n = n + p; p = p * w end
  return n
end

local sink = 0

local function measure(warm, runs, f)
  for _ = 1, warm do f() end
  local t = {}
  for _ = 1, runs do
    local a = os.clock()
    f()
    t[#t + 1] = (os.clock() - a) * 1000
  end
  table.sort(t)
  local s = 0
  for _, x in ipairs(t) do s = s + x end
  return t[1], t[math.floor(#t / 2) + 1], s / #t
end

local tree = build(W, D, 0)
local nodes = nodecount(W, D)
local treeA = build(W, D, 1)
local treeB = build(W, D, 2)
local mlist = setmetatable({ treeA, treeB }, { __jsontype = "array" })
local segs = {}
for i = 1, D do segs[i] = "k0" end
local path = table.concat(segs, ".")
local cb = function(_k, v, _p, pth) sink = sink + #pth; return v end

local specs = {
  { "clone", nodes, function() struct.clone(tree); sink = sink + 1 end },
  { "walk", nodes, function() struct.walk(tree, cb) end },
  { "merge", nodes, function() struct.merge(mlist); sink = sink + 1 end },
  { "stringify", nodes, function() sink = sink + #struct.stringify(tree) end },
  { "getpath", GP, function()
    local a = 0
    for _ = 1, GP do if struct.getpath(tree, path) == 0 then a = a + 1 end end
    sink = sink + a
  end },
}

local parts = {}
for _, sp in ipairs(specs) do
  local mn, md, mean = measure(WARM, RUNS, sp[3])
  parts[#parts + 1] = string.format(
    '{"op":"%s","runs":%d,"unit_count":%d,"min_ms":%.6g,"median_ms":%.6g,"mean_ms":%.6g}',
    sp[1], RUNS, sp[2], mn, md, mean)
end

io.stderr:write("lua: sink=" .. sink .. "\n")
io.write(string.format(
  '{"lang":"lua","runtime":"%s","nodes":%d,"params":{"width":%d,"depth":%d,"warmup":%d,"runs":%d,"getpath_iters":%d},"ops":[%s]}\n',
  _VERSION:lower():gsub(" ", " "), nodes, W, D, WARM, RUNS, GP, table.concat(parts, ",")))
